src/neo4j_genai/retrievers/text2cypher.py

```diff
@@ -0,0 +1,109 @@
+    """
+    Allows for the retrieval of records from a Neo4j database using natural language.
+    Converts a user's natural language query to a Cypher query using an LLM,
+    then retrieves records from a Neo4j database using the generated Cypher query
+    """
+
+    def __init__(
+        self, driver: neo4j.Driver, llm: LLM, neo4j_schema: Optional[str] = None
+    ) -> None:
+        try:
+            driver_model = Neo4jDriverModel(driver=driver)
+            llm_model = LLMModel(llm=llm)
+            neo4j_schema_model = (
+                Neo4jSchemaModel(neo4j_schema=neo4j_schema) if neo4j_schema else None
+            )
+            validated_data = Text2CypherRetrieverModel(
+                driver_model=driver_model,
+                llm_model=llm_model,
+                neo4j_schema_model=neo4j_schema_model,
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        super().__init__(validated_data.driver_model.driver)
+        self.llm = validated_data.llm_model.llm
+        self.neo4j_schema = (
+            validated_data.neo4j_schema_model.neo4j_schema
+            if validated_data.neo4j_schema_model
+            else get_schema(validated_data.driver_model.driver)
+        )
+
+    def search(
+        self, query_text: str, examples: Optional[list[str]] = None
+    ) -> list[neo4j.Record]:
+        """Converts query_text to a Cypher query using an LLM.
+           Retrieve records from a Neo4j database using the generated Cypher query.
+
+        Args:
+            query_text (str): The natural language query used to search the Neo4j database.
+            examples (Optional[list[str], optional): Optional user input/query pairs for the LLM to use as examples.
+
+        Raises:
+            ValidationError: If validation of the input arguments fail.
+
+        Returns:
+            list[neo4j.Record]: The results of the search query.
+        """
+        try:
+            validated_data = Text2CypherSearchModel(
+                query_text=query_text, examples=examples
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        prompt = TEXT2CYPHER_PROMPT.format(
+            schema=self.neo4j_schema,
+            examples="\n".join(validated_data.examples)
+            if validated_data.examples
+            else "",
+            input=validated_data.query_text,
+        )
+        logger.debug("Text2CypherRetriever prompt: %s", prompt)
+
+        try:
+            t2c_query = self.llm.invoke(prompt)
+            logger.debug("Text2CypherRetriever Cypher query: %s", t2c_query)
+            records, _, _ = self.driver.execute_query(query_=t2c_query)
+        except CypherSyntaxError as e:
+            logger.error("Cypher query generation failed: %s", e.message)
+            raise RuntimeError(f"Cypher query generation failed: {e.message}")
+
+        return records
```

Comment:
It seems useful to return the generated query here, in addition to the records (would be useful to show in UI:s etc).

Right now we don't have a shared return type from our retrievers, but we would need to have that once we want to add the `Neo4jRAG` instance to handle retrieval.

We have alternatives though, maybe there's a different return type depending on a flag that's passed in `include_metadata` or something.
I don't know, please think a bit about this both for this local code but also for the larger story.

Reply:
In the RAG-eval code, we used this model which is returned by all retrievers:
```python
class RetrieverResult(BaseModel):
    documents: list[Document]
    execution_metadata: Optional[dict] = None
```
The generated cypher query (if any), is stored in the `execution_metadata` dict.

Reply:
What sort of metadata will need to be returned when using the other retrievers? LangChain retrievers return a list of `Document` objects (see [here](https://api.python.langchain.com/en/latest/documents/langchain_core.documents.base.Document.html#langchain_core.documents.base.Document)) which allow you to attach arbitrary pieces of metadata to each document. We could do something similar here? Alternatively we return one `Metadata` object with each search? I guess it depends if we expect different metadata for each record. The first approach would be more extensible but for this specific retriever would mean attaching the query to each record which would be a tad wasteful.